src/Mods.jl

```diff
@@ -49,224 +49,130 @@ julia> value(a)
+Base.promote_rule(::Type{Mod{M,T1}}, ::Type{Mod{M,T2}}) where {M,T1,T2<:Number} = Mod{M,promote_type(T1, T2)}
+Base.promote_rule(::Type{Mod{M,T1}}, ::Type{T2}) where {M,T1,T2<:Number} = Mod{M,promote_type(T1, T2)}
+Base.promote_rule(::Type{Mod{M,T1}}, ::Type{Rational{T2}}) where {M,T1,T2} = Mod{M,promote_type(T1, T2)}
 
 # Operations with rational numbers  
-
 Mod{N}(k::Rational) where N = Mod{N}(numerator(k))/Mod{N}(denominator(k))
-
-function +(x::Mod{N}, k::Rational) where N
-    return x + Mod{N}(k)
-end
-(+)(k::Rational,x::Mod) = x+k
-
-(-)(x::Mod,k::Rational) = x + (-k)
-(-)(k::Rational,x::Mod) = k + (-x)
-
-function (*)(x::Mod{N},k::Rational) where N
-    return x * Mod{N}(k)
-end
-(*)(k::Rational,x::Mod) = x*k
-
-function (/)(x::Mod,k::Rational)
-    return x * (1/k)
-end
-(/)(k::Rational,x::Mod) = k * inv(x)
-
-(//)(x::Mod,k::Rational) = x/k
-(//)(k::Rational,x::Mod) = k/x
-
-
-
-
-
-
-# Comparison with Integers
-
-isequal(x::Mod{M}, k::Integer) where M = mod(k,M) == x.val
-isequal(k::Integer, x::Mod) = isequal(x,k)
-(==)(x::Mod, k::Integer) = isequal(x,k)
-(==)(k::Integer, x::Mod) = isequal(x,k)
-
-# Comparisons with Rationals
-function isequal(x::Mod{N}, k::Rational) where N
-    return x == Mod{N}(k)
-end
-isequal(k::Rational,x::Mod) = isequal(x,k)
-(==)(x::Mod, k::Rational) = isequal(x,k)
-(==)(k::Rational, x::Mod) = isequal(x,k)
-
-
-
+Mod{N,T}(k::Rational{T2}) where {N,T,T2} = Mod{N,T}(numerator(k))/Mod{N,T}(denominator(k))
 
 # Random
-
-rand(::Type{Mod{N}}) where N = Mod{N}(rand(Int))
-rand(::Type{Mod{N}},dims::Integer...) where N = Mod{N}.(rand(Int,dims...))
-
-
+rand(::Type{Mod{N}}, args::Integer...) where {N} = rand(Mod{N,Int}, args...)
+rand(::Type{Mod{N,T}}) where {N,T} = Mod{N}(rand(T))
+rand(::Type{Mod{N,T}},dims::Integer...) where {N,T} = Mod{N}.(rand(T,dims...))
 
 # Chinese remainder theorem functions
-
-# private helper function
-function CRT_work(x::Mod{n}, y::Mod{m}) where {n,m}
-    # n = x.mod
-    # m = y.mod
-    if gcd(n,m) != 1
-        error("Moduli must be pairwise relatively prime")
-    end
-
-    a = x.val
-    b = y.val
-
-    k = inv(Mod(n,m)) * (b-a)
-
-    z = a + k.val*n
-
-    return Mod(z, n*m)
-end
-
-# public interface
 """
-`CRT(m1,m2,...)`: Chinese Remainder Theorem
+`CRT(m1, m2,...)`: Chinese Remainder Theorem
+
 ```
-julia> CRT( Mod(4,11), Mod(8,14) )
-Mods.Mod(92,154)
+julia> CRT(Mod{11}(4), Mod{14}(814))
+92
 
 julia> 92%11
 4
 
 julia> 92%14
 8
+
+julia> CRT(BigInt, Mod{9223372036854775783}(9223372036854775782), Mod{9223372036854775643}(9223372036854775642))
 ```
 """
-function CRT(mtuple::Mod...)
-    n = length(mtuple)
-    if n == 0
-        return 1
-    end
-
-    result::Mod = mtuple[1]
-
-    for k=2:n
-        result = CRT_work(result,mtuple[k])
-    end
+function CRT(remainders, primes)
+    length(remainders) == length(primes) || error("size mismatch")
+    isempty(remainders) && return 1
```

Comment:
Type stability issue: The returned `1` has not always the same type as the final return value (`typeof(M)`).

Reply:
I throwed an error instead. It does not make sense to feed an empty vector into this function.